CriteoPublisherSdk/Sources/CR_BidManager.m

```diff
@@ -179,6 +180,47 @@ - (void)prefetchBidsForAdUnits:(CR_CacheAdUnitArray *)adUnits {
          }];
 }
 
+- (void)fetchLiveBidForAdUnit:(CR_CacheAdUnit *)adUnit
+           bidResponseHandler:(CR_BidResponseHandler)responseHandler {
+  [self fetchLiveBidForAdUnit:adUnit
+           bidResponseHandler:responseHandler
+                   timeBudget:CRITEO_DEFAULT_LIVE_BID_TIME_BUDGET_IN_SECONDS];
+}
+
+- (void)fetchLiveBidForAdUnit:(CR_CacheAdUnit *)adUnit
+           bidResponseHandler:(CR_BidResponseHandler)responseHandler
+                   timeBudget:(NSTimeInterval)timeBudget {
+  [self.threadManager dispatchAsyncOnGlobalQueueWithTimeout:timeBudget
+      operationHandler:^void(void (^completionHandler)(dispatchWithTimeoutHandler)) {
+        [self fetchBidsForAdUnits:@[ adUnit ]
+               cdbResponseHandler:^(CR_CdbResponse *cdbResponse) {
+                 completionHandler(^(BOOL handled) {
+                   [self.threadManager dispatchAsyncOnMainQueue:^{
+                     if (!handled) {
+                       NSAssert(cdbResponse.cdbBids.count <= 1,
+                                @"During a live request, only one bid will be fetched at a time.");
+                       if (cdbResponse.cdbBids.count == 1) {
+                         CR_CdbBid *bid = cdbResponse.cdbBids[0];
+                         responseHandler(bid);
```

Comment:
It seems that you don't apply the silence/expired/no bid/user silence logic here.
If your response handler is only accepting nil or valid bids, then you need to validate the response:
- bid is valid if cpm > 0 and ttl >= 0
- bid is silence if cpm == 0 and ttl > 0
- bid is no bid if cpm <= 0 or ttl < 0

Reply:
This logic partially in `BidManager.getBid()` and `CacheManager.setBid()`, but clearly missing there, will refactor code so it applies consistently to these two code paths. Thanks. I guess CSM needs some rework as well.

Reply:
Fixed by:
- [livebid][BidManager] Handle slot level silent 0f6139e
- [livebid][BidManager] Handle valid and expired bids 877c68a
Still some more work to come, especially regarding CSM, but I would say FAR.
Also there is no check yet in cache before fetching to handle silent slot, FAR as well if that is fine with you.